File: iosApp/Sources/Service/ConfigService.swift
import Foundation

enum AppPaths {
    static var baseDir: URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("harnest", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static var harnessDir: URL { baseDir.appendingPathComponent("harness", isDirectory: true) }
}

/// AI provider config — mirrors ConfigService.kt / ConfigService.ets (EMM-compatible).
/// Persisted at ApplicationSupport/harnest/api_config.json.
final class ConfigService {

    private static let file = AppPaths.baseDir.appendingPathComponent("api_config.json")

    private static let shared = ConfigService()

    static func get() -> ConfigService { shared }

    private let lock = NSLock()
    private var cached: [String: Any]?

    private func loadLocked() -> [String: Any] {
        if let c = cached { return c }
        var read: [String: Any]? = nil
        if let data = try? Data(contentsOf: Self.file) {
            read = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        }
        if var read = read, read["configs"] is [String: Any] {
            if (read["presetMigration"] as? String) != "2026-08b" {
                // 存量配置一次性升级：旧预设模型列表换为官方现役列表（定制过的补齐缺失项）
                Self.migratePresets(&read)
                read["presetMigration"] = "2026-08b"
                cached = read
                saveLocked(read)
                return read
            }
            cached = read
            return read
        }
        var configs: [String: Any] = [:]
        for p in Providers.all {
            guard let meta = Providers.metaOf(p) else { continue }
            configs[p] = [
                "provider": p,
                "apiKey": "",
                "baseUrl": meta.baseUrl,
                "models": meta.models,
                "defaultModel": meta.defaultModel,
                "enabled": false,
            ] as [String: Any]
        }
        let fresh: [String: Any] = [
            "version": 1,
            "configs": configs,
            "lastProvider": Providers.deepseek,
            "lastModel": Providers.metaOf(Providers.deepseek)?.defaultModel ?? "",
        ]
        cached = fresh
        saveLocked(fresh)
        return fresh
    }

    private func saveLocked(_ cfg: [String: Any]) {
        // Swift 字典为值类型：写路径必须回写缓存，否则 load() 仍返回旧快照
        // （Kotlin JSONObject 为引用类型，原地 put 即同步；Swift 需显式回写）
        cached = cfg
        guard let data = try? JSONSerialization.data(withJSONObject: cfg, options: [.prettyPrinted, .sortedKeys]) else { return }
        try? data.write(to: Self.file, options: .atomic)
    }

    /// 预设刷新迁移（2026-08）：官方模型 ID 大面积退役后，存量配置里
    /// 全部落在旧预设内的模型列表整体替换为新预设；定制过的列表保留用户条目，
    /// 仅把缺失的现役预设补入（保证 kimi-k3 等新 ID 始终可选）；
    /// defaultModel/lastModel 若已退役一并纠正。
    private static func migratePresets(_ cfg: inout [String: Any]) {
        guard var configs = cfg["configs"] as? [String: Any] else { return }
        for p in Providers.all {
            guard var item = configs[p] as? [String: Any],
                  let legacy = Providers.legacyPresetModels[p],
                  let meta = Providers.metaOf(p) else { continue }
            let models = (item["models"] as? [Any])?.compactMap { $0 as? String } ?? []
            if models.allSatisfy({ legacy.contains($0) }) {
                item["models"] = meta.models
            } else {
                var merged = models
                for m in meta.models where !merged.contains(m) {
                    merged.append(m)
                }
                item["models"] = merged
            }
            if let def = item["defaultModel"] as? String, legacy.contains(def) {
                item["defaultModel"] = meta.defaultModel
            }
            configs[p] = item
        }
        cfg["configs"] = configs
        let lastProvider = cfg["lastProvider"] as? String ?? ""
        if let legacyLast = Providers.legacyPresetModels[lastProvider],
           let metaLast = Providers.metaOf(lastProvider),
           let lastModel = cfg["lastModel"] as? String,
           legacyLast.contains(lastModel) {
            cfg["lastModel"] = metaLast.defaultModel
        }
    }

    func load() -> [String: Any] {
        lock.lock(); defer { lock.unlock() }
        return loadLocked()
    }

    func save() {
        lock.lock(); defer { lock.unlock() }
        if let c = cached { saveLocked(c) }
    }

    func getConfig(_ provider: String) -> [String: Any] {
        let c = load()
        if let item = (c["configs"] as? [String: Any])?[provider] as? [String: Any] { return item }
        let meta = Providers.metaOf(provider)
        return [
            "provider": provider,
            "apiKey": "",
            "baseUrl": meta?.baseUrl ?? "",
            "models": meta?.models ?? [],
            "defaultModel": meta?.defaultModel ?? "",
            "enabled": false,
        ] as [String: Any]
    }

    /// Save one provider entry; non-empty apiKey == enabled.
    func setConfig(provider: String, apiKey: String, baseUrl: String, models: [String], defaultModel: String, maxTokens: Int? = nil) {
        lock.lock(); defer { lock.unlock() }
        var c = loadLocked()
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        let effectiveBase = baseUrl.isEmpty ? (Providers.metaOf(provider)?.baseUrl ?? "") : baseUrl
        var entry: [String: Any] = [
            "provider": provider,
            "apiKey": trimmedKey,
            "baseUrl": effectiveBase,
            "models": models.isEmpty ? [defaultModel] : models,
            "defaultModel": defaultModel.isEmpty ? (models.first ?? "") : defaultModel,
            "enabled": !trimmedKey.isEmpty,
        ]
        if let mt = maxTokens, mt > 0 {
            entry["maxTokens"] = mt
        }
        var configs = (c["configs"] as? [String: Any]) ?? [:]
        configs[provider] = entry
        c["configs"] = configs
        saveLocked(c)
    }

    func setLastSelection(provider: String, model: String, effort: String? = nil) {
        lock.lock(); defer { lock.unlock() }
        var c = loadLocked()
        c["lastProvider"] = provider
        c["lastModel"] = model
        if let v = effort { c["lastEffort"] = v } else { c.removeValue(forKey: "lastEffort") }
        saveLocked(c)
    }

    /// 最近一次发送所用的思考模式；nil = 默认。
    func getDefaultEffort() -> String? {
        let v = load()["lastEffort"] as? String ?? ""
        guard !v.isEmpty else { return nil }
        return ReasoningEfforts.isValid(v) ? v : nil
    }

    func hasUsableConfig() -> Bool { !listUsableProviders().isEmpty }

    /// Provider entries with a non-empty apiKey, in catalog order.
    func listUsableProviders() -> [[String: Any]] {
        let configs = (load()["configs"] as? [String: Any]) ?? [:]
        var out: [[String: Any]] = []
        for p in Providers.all {
            guard let item = configs[p] as? [String: Any] else { continue }
            let key = (item["apiKey"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if !key.isEmpty { out.append(item) }
        }
        return out
    }

    /// Default pick: last selection if still enabled, else first usable.
    func getDefaultSelection() -> (String, String) {
        let c = load()
        let lastProvider = c["lastProvider"] as? String ?? ""
        if !lastProvider.isEmpty {
            let item = (c["configs"] as? [String: Any])?[lastProvider] as? [String: Any]
            if let item = item, item["enabled"] as? Bool == true {
                let models = jsonStrings(item["models"])
                let lastModel = c["lastModel"] as? String ?? ""
                let model = (models.contains(lastModel) && !lastModel.isEmpty)
                    ? lastModel
                    : (item["defaultModel"] as? String ?? "")
                return (lastProvider, model)
            }
        }
        let usable = listUsableProviders()
        if let first = usable.first {
            return (first["provider"] as? String ?? "", first["defaultModel"] as? String ?? "")
        }
        return (Providers.deepseek, Providers.metaOf(Providers.deepseek)?.defaultModel ?? "")
    }

    /// Kernel init payload: {cwd, providers[], defaultProvider, defaultModel} — nil when unconfigured.
    func buildEngineConfig(cwd: String) -> [String: Any]? {
        let usable = listUsableProviders()
        if usable.isEmpty { return nil }
        let def = getDefaultSelection()
        var profiles: [[String: Any]] = []
        for item in usable {
            var models = jsonStrings(item["models"])
            if models.isEmpty { models = [item["defaultModel"] as? String ?? ""] }
            var profile: [String: Any] = [
                "provider": item["provider"] as? String ?? "",
                "baseUrl": item["baseUrl"] as? String ?? "",
                "apiKey": item["apiKey"] as? String ?? "",
                "models": models.map { ["id": $0] as [String: Any] },
            ]
            if let mt = item["maxTokens"] as? Int, mt > 0 {
                profile["maxTokens"] = mt
            }
            profiles.append(profile)
        }
        return [
            "cwd": cwd,
            "providers": profiles,
            "defaultProvider": def.0,
            "defaultModel": def.1,
        ]
    }

    /// Import EMM/harness_model_config JSON (from clipboard or file). Returns (imported, skipped).
    func importConfigJson(_ jsonStr: String) -> (Int, [String]) {
        lock.lock(); defer { lock.unlock() }
        guard let parsed = (try? JSONSerialization.jsonObject(with: Data(jsonStr.utf8))) as? [String: Any] else {
            return (0, [])
        }
        let type = parsed["type"] as? String ?? ""
        if type != "harness_model_config" && type != "emm_model_config" { return (0, []) }
        guard let arr = parsed["configs"] as? [[String: Any]] else { return (0, []) }
        var c = loadLocked()
        var configs = (c["configs"] as? [String: Any]) ?? [:]
        var imported = 0
        var skipped: [String] = []
        for obj in arr {
            let providerRaw = obj["provider"] as? String ?? ""
            let apiKey = (obj["apiKey"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if providerRaw.isEmpty { continue }
            let provider = Providers.importAlias[providerRaw] ?? providerRaw
            if Providers.importUnsupported.contains(provider) {
                if !skipped.contains(provider) { skipped.append(provider) }
                continue
            }
            guard let meta = Providers.metaOf(provider) else { continue }
            if apiKey.isEmpty || apiKey.count > 500 { continue }
            var models = jsonStrings(obj["models"])
            if models.isEmpty { models = meta.models }
            var defaultModel = (obj["defaultModel"] as? String) ?? (obj["model"] as? String) ?? ""
            if defaultModel.isEmpty { defaultModel = meta.defaultModel }
            if !models.contains(defaultModel) { models.append(defaultModel) }
            var baseUrl = obj["baseUrl"] as? String ?? ""
            if baseUrl.isEmpty { baseUrl = meta.baseUrl }
            if provider == Providers.gemini && !baseUrl.contains("/openai") { baseUrl = meta.baseUrl }
            var entry: [String: Any] = [
                "provider": provider,
                "apiKey": apiKey,
                "baseUrl": baseUrl,
                "models": models,
                "defaultModel": defaultModel,
                "enabled": true,
            ]
            if let mt = obj["maxTokens"] as? Int, mt > 0 {
                entry["maxTokens"] = mt
            }
            if let existing = configs[provider] as? [String: Any],
               existing["apiKey"] as? String == apiKey,
               existing["defaultModel"] as? String == defaultModel {
                continue
            }
            configs[provider] = entry
            imported += 1
        }
        if imported > 0 {
            c["configs"] = configs
            saveLocked(c)
        }
        return (imported, skipped)
    }

    /// Export EMM-compatible model config JSON (only providers with an apiKey).
    func exportConfigJson() -> String {
        let c = load()
        var configs: [[String: Any]] = []
        for p in Providers.all {
            guard let item = (c["configs"] as? [String: Any])?[p] as? [String: Any] else { continue }
            let apiKey = (item["apiKey"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if apiKey.isEmpty { continue }
            var exported: [String: Any] = [
                "provider": item["provider"] as? String ?? p,
                "apiKey": apiKey,
                "model": item["defaultModel"] as? String ?? "",
                "baseUrl": item["baseUrl"] as? String ?? "",
                "models": item["models"] ?? [],
                "defaultModel": item["defaultModel"] as? String ?? "",
            ]
            if let mt = item["maxTokens"] as? Int, mt > 0 {
                exported["maxTokens"] = mt
            }
            configs.append(exported)
        }
        let out: [String: Any] = [
            "version": 1,
            "type": "emm_model_config",
            "exportTime": Int(Date().timeIntervalSince1970 * 1000),
            "configs": configs,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: out, options: [.prettyPrinted, .sortedKeys]) else {
            return "{}"
        }
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    private func jsonStrings(_ any: Any?) -> [String] {
        guard let arr = any as? [Any] else { return [] }
        return arr.compactMap { ($0 as? String)?.isEmpty == false ? $0 as? String : nil }
    }
}

File: androidApp/app/src/main/java/com/harnest/app/service/ConfigService.kt
package com.harnest.app.service

import android.content.Context
import android.util.Log
import com.harnest.app.common.Providers
import com.harnest.app.common.ReasoningEfforts
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * AI provider config — mirrors harmonyApp ConfigService.ets (EMM-compatible).
 * Persisted at filesDir/api_config.json as {version, configs, lastProvider, lastModel}.
 */
class ConfigService private constructor(private val context: Context) {

    companion object {
        private const val TAG = "ConfigService"
        private const val FILE = "api_config.json"

        @Volatile
        private var inst: ConfigService? = null

        fun get(context: Context): ConfigService =
            inst ?: synchronized(this) {
                inst ?: ConfigService(context.applicationContext).also { inst = it }
            }
    }

    private var cached: JSONObject? = null

    private fun file(): File = File(context.filesDir, FILE)

    @Synchronized
    fun load(): JSONObject {
        cached?.let { return it }
        val read: JSONObject? = try {
            if (file().exists()) JSONObject(file().readText()) else null
        } catch (e: Throwable) {
            Log.e(TAG, "config read failed", e); null
        }
        if (read != null && read.has("configs")) {
            if (read.optString("presetMigration") != "2026-08b") {
                // 存量配置一次性升级：旧预设模型列表换为官方现役列表（定制过的补齐缺失项）
                migratePresets(read)
                read.put("presetMigration", "2026-08b")
                cached = read
                saveLocked(read)
                Log.i(TAG, "config migrated to 2026-08b presets")
                return read
            }
            cached = read
            return read
        }
        val configs = JSONObject()
        for (p in Providers.ALL) {
            val meta = Providers.metaOf(p)!!
            configs.put(p, JSONObject()
                .put("provider", p)
                .put("apiKey", "")
                .put("baseUrl", meta.baseUrl)
                .put("models", JSONArray(meta.models))
                .put("defaultModel", meta.defaultModel)
                .put("enabled", false))
        }
        val fresh = JSONObject()
            .put("version", 1)
            .put("configs", configs)
            .put("lastProvider", Providers.DEEPSEEK)
            .put("lastModel", Providers.metaOf(Providers.DEEPSEEK)!!.defaultModel)
        cached = fresh
        saveLocked(fresh)
        return fresh
    }

    @Synchronized
    fun save() {
        cached?.let { saveLocked(it) }
    }

    private fun saveLocked(cfg: JSONObject) {
        try {
            file().writeText(cfg.toString(2))
        } catch (e: Throwable) {
            Log.e(TAG, "config save failed", e)
        }
    }

    /**
     * 预设刷新迁移（2026-08）：官方模型 ID 大面积退役后，存量配置里
     * 全部落在旧预设内的模型列表整体替换为新预设；定制过的列表保留用户条目，
     * 仅把缺失的现役预设补入（保证 kimi-k3 等新 ID 始终可选）；
     * defaultModel/lastModel 若已退役一并纠正。
     */
    private fun migratePresets(cfg: JSONObject) {
        val configs = cfg.optJSONObject("configs") ?: return
        for (p in Providers.ALL) {
            val item = configs.optJSONObject(p) ?: continue
            val legacy = Providers.LEGACY_PRESET_MODELS[p] ?: continue
            val meta = Providers.metaOf(p) ?: continue
            val models = item.optJSONArray("models") ?: continue
            val list = (0 until models.length()).mapNotNull { models.optString(it) }
            if (list.all { legacy.contains(it) }) {
                item.put("models", JSONArray(meta.models))
            } else {
                val merged = JSONArray(list)
                for (m in meta.models) {
                    if (!list.contains(m)) merged.put(m)
                }
                item.put("models", merged)
            }
            if (legacy.contains(item.optString("defaultModel"))) {
                item.put("defaultModel", meta.defaultModel)
            }
        }
        val legacyLast = Providers.LEGACY_PRESET_MODELS[cfg.optString("lastProvider")]
        val metaLast = Providers.metaOf(cfg.optString("lastProvider"))
        if (legacyLast != null && metaLast != null && legacyLast.contains(cfg.optString("lastModel"))) {
            cfg.put("lastModel", metaLast.defaultModel)
        }
    }

    fun getConfig(provider: String): JSONObject {
        val c = load()
        c.optJSONObject("configs")?.optJSONObject(provider)?.let { return it }
        val meta = Providers.metaOf(provider)
        val models: List<String> = meta?.models ?: emptyList()
        return JSONObject()
            .put("provider", provider)
            .put("apiKey", "")
            .put("baseUrl", meta?.baseUrl ?: "")
            .put("models", JSONArray(models))
            .put("defaultModel", meta?.defaultModel ?: "")
            .put("enabled", false)
    }

    /** Save one provider entry; non-empty apiKey == enabled. */
    @Synchronized
    fun setConfig(
        provider: String,
        apiKey: String,
        baseUrl: String,
        models: List<String>,
        defaultModel: String,
        maxTokens: Int? = null,
    ) {
        val c = load()
        val trimmedKey = apiKey.trim()
        if (baseUrl.isBlank() && provider != Providers.CUSTOM) {
            // keep preset when caller passes blank
        }
        val entry = JSONObject()
            .put("provider", provider)
            .put("apiKey", trimmedKey)
            .put("baseUrl", baseUrl.ifBlank { Providers.metaOf(provider)?.baseUrl ?: "" })
            .put("models", JSONArray(models.ifEmpty { listOf(defaultModel) }))
            .put("defaultModel", defaultModel.ifBlank { models.firstOrNull() ?: "" })
            .put("enabled", trimmedKey.isNotEmpty())
        if (maxTokens != null && maxTokens > 0) {
            entry.put("maxTokens", maxTokens)
        }
        c.optJSONObject("configs")?.put(provider, entry) ?: c.put("configs", JSONObject().put(provider, entry))
        saveLocked(c)
    }

    @Synchronized
    fun setLastSelection(provider: String, model: String, effort: String? = null) {
        val c = load()
        c.put("lastProvider", provider).put("lastModel", model)
        if (effort != null) c.put("lastEffort", effort) else c.remove("lastEffort")
        saveLocked(c)
    }

    /** 最近一次发送所用的思考模式；null = 默认。 */
    fun getDefaultEffort(): String? {
        val v = load().optString("lastEffort", "")
        return if (ReasoningEfforts.isValid(v.ifEmpty { null })) v.ifEmpty { null } else null
    }

    fun hasUsableConfig(): Boolean = listUsableProviders().isNotEmpty()

    /** Provider entries with a non-empty apiKey. */
    fun listUsableProviders(): List<JSONObject> {
        val out = ArrayList<JSONObject>()
        val configs = load().optJSONObject("configs") ?: return out
        for (p in Providers.ALL) {
            val item = configs.optJSONObject(p) ?: continue
            if (item.optString("apiKey").trim().isNotEmpty()) out.add(item)
        }
        return out
    }

    /** Default pick: last selection if still enabled, else first usable. */
    fun getDefaultSelection(): Pair<String, String> {
        val c = load()
        val lastProvider = c.optString("lastProvider", "")
        if (lastProvider.isNotEmpty()) {
            val item = c.optJSONObject("configs")?.optJSONObject(lastProvider)
            if (item != null && item.optBoolean("enabled", false)) {
                val models = jsonStrings(item.optJSONArray("models"))
                val lastModel = c.optString("lastModel", "")
                val model = if (lastModel.isNotEmpty() && models.contains(lastModel)) lastModel
                else item.optString("defaultModel")
                return Pair(lastProvider, model)
            }
        }
        val usable = listUsableProviders()
        if (usable.isNotEmpty()) {
            return Pair(usable[0].optString("provider"), usable[0].optString("defaultModel"))
        }
        return Pair(Providers.DEEPSEEK, Providers.metaOf(Providers.DEEPSEEK)!!.defaultModel)
    }

    /** Kernel init payload: {cwd, providers[], defaultProvider, defaultModel} — null when unconfigured. */
    fun buildEngineConfig(cwd: String): JSONObject? {
        val usable = listUsableProviders()
        if (usable.isEmpty()) return null
        val def = getDefaultSelection()
        val profiles = JSONArray()
        for (item in usable) {
            val models = jsonStrings(item.optJSONArray("models"))
                .ifEmpty { listOf(item.optString("defaultModel")) }
            val modelInputs = JSONArray()
            for (m in models) modelInputs.put(JSONObject().put("id", m))
            val profile = JSONObject()
                .put("provider", item.optString("provider"))
                .put("baseUrl", item.optString("baseUrl"))
                .put("apiKey", item.optString("apiKey"))
                .put("models", modelInputs)
            val mt = item.optInt("maxTokens", 0)
            if (mt > 0) profile.put("maxTokens", mt)
            profiles.put(profile)
        }
        return JSONObject()
            .put("cwd", cwd)
            .put("providers", profiles)
            .put("defaultProvider", def.first)
            .put("defaultModel", def.second)
    }

    /** Import EMM/harness_model_config JSON (from clipboard or file). Returns imported count + skipped providers. */
    @Synchronized
    fun importConfigJson(jsonStr: String): Pair<Int, List<String>> {
        try {
            val parsed = JSONObject(jsonStr)
            val type = parsed.optString("type", "")
            if (type != "harness_model_config" && type != "emm_model_config") return Pair(0, emptyList())
            val arr = parsed.optJSONArray("configs") ?: return Pair(0, emptyList())
            val c = load()
            val configs = c.optJSONObject("configs") ?: JSONObject().also { c.put("configs", it) }
            var imported = 0
            val skipped = ArrayList<String>()
            for (i in 0 until arr.length()) {
                val obj = arr.optJSONObject(i) ?: continue
                val providerRaw = obj.optString("provider", "")
                val apiKey = obj.optString("apiKey", "").trim()
                if (providerRaw.isEmpty()) continue
                val provider = Providers.IMPORT_ALIAS[providerRaw] ?: providerRaw
                if (Providers.IMPORT_UNSUPPORTED.contains(provider)) {
                    if (!skipped.contains(provider)) skipped.add(provider)
                    continue
                }
                val meta = Providers.metaOf(provider) ?: continue
                if (apiKey.isEmpty() || apiKey.length > 500) continue
                var models = jsonStrings(obj.optJSONArray("models")).ifEmpty { meta.models }
                var defaultModel = obj.optString("defaultModel", obj.optString("model", ""))
                if (defaultModel.isEmpty()) defaultModel = meta.defaultModel
                if (!models.contains(defaultModel)) models = models + defaultModel
                var baseUrl = obj.optString("baseUrl", "")
                if (baseUrl.isEmpty()) baseUrl = meta.baseUrl
                if (provider == Providers.GEMINI && !baseUrl.contains("/openai")) baseUrl = meta.baseUrl
                val entry = JSONObject()
                    .put("provider", provider)
                    .put("apiKey", apiKey)
                    .put("baseUrl", baseUrl)
                    .put("models", JSONArray(models))
                    .put("defaultModel", defaultModel)
                    .put("enabled", true)
                val mt = obj.optInt("maxTokens", 0)
                if (mt > 0) entry.put("maxTokens", mt)
                val existing = configs.optJSONObject(provider)
                if (existing != null && existing.optString("apiKey") == apiKey
                    && existing.optString("defaultModel") == defaultModel) continue
                configs.put(provider, entry)
                imported++
            }
            if (imported > 0) saveLocked(c)
            return Pair(imported, skipped)
        } catch (e: Throwable) {
            Log.e(TAG, "import failed", e)
            return Pair(0, emptyList())
        }
    }

    /** Export EMM-compatible model config JSON (only providers with an apiKey). */
    @Synchronized
    fun exportConfigJson(): String {
        val c = load()
        val configs = JSONArray()
        for (p in Providers.ALL) {
            val item = c.optJSONObject("configs")?.optJSONObject(p) ?: continue
            val apiKey = item.optString("apiKey").trim()
            if (apiKey.isEmpty()) continue
            val exported = JSONObject()
                .put("provider", item.optString("provider", p))
                .put("apiKey", apiKey)
                .put("model", item.optString("defaultModel"))
                .put("baseUrl", item.optString("baseUrl"))
                .put("models", item.optJSONArray("models") ?: JSONArray())
                .put("defaultModel", item.optString("defaultModel"))
            val mt = item.optInt("maxTokens", 0)
            if (mt > 0) exported.put("maxTokens", mt)
            configs.put(exported)
        }
        return JSONObject()
            .put("version", 1)
            .put("type", "emm_model_config")
            .put("exportTime", System.currentTimeMillis())
            .put("configs", configs)
            .toString(2)
    }

    private fun jsonStrings(arr: JSONArray?): List<String> {
        if (arr == null) return emptyList()
        val out = ArrayList<String>(arr.length())
        for (i in 0 until arr.length()) {
            val s = arr.optString(i, "")
            if (s.isNotEmpty()) out.add(s)
        }
        return out
    }
}
